File: 2021/2021W35/PreppindataW35.py
import pandas as pd
import re
import numpy as np

xlsx = pd.ExcelFile(r"F:\Data\Pictures Input.xlsx")

def analyze_size(size_str):
    side_a = float(re.search('^(\d+)', size_str).group(1))
    side_b = float(re.search('\D(\d+)\D', size_str).group(1) if re.search('\D(\d+)\D', size_str) else side_a)
    side_all = [side_a, side_b] if re.search('(cm)', size_str) else [side_a*2.54, side_b*2.54]
    side_all.sort()
    
    return side_all

picture = pd.read_excel(xlsx,'Pictures')
picture['Size Side'] = picture['Size'].apply(lambda x: analyze_size(x))
picture['Min Side'] = picture['Size Side'].apply(lambda x: x[0])
picture['Max Side'] = picture['Size Side'].apply(lambda x: x[1])

frame = pd.read_excel(xlsx,'Frames')
frame.columns = ['Frame']
frame['Frame Size Side'] = frame['Frame'].apply(lambda x: analyze_size(x))

final = pd.merge(picture, frame, how='cross')
final['Fit'] = final.apply(lambda x: x['Frame Size Side'][0]>=x['Size Side'][0] and x['Frame Size Side'][1]>=x['Size Side'][1], axis=1)
final = final[final['Fit']].copy()
final['Excess Area'] = final.apply(lambda x: np.prod(x['Frame Size Side'])-np.prod(x['Size Side']), axis=1)
final['Min Excess Area'] = final['Excess Area'].groupby(final['Picture']).transform('min')

final = final[final['Excess Area']==final['Min Excess Area']][['Picture', 'Frame', 'Max Side', 'Min Side']]

File: 2021/2021W35/preppindataW35.R
library(readxl)
library(dplyr)
library(tidyr)
library(stringr)
library(fuzzyjoin)

input <- "F:/Data/Pictures Input.xlsx"

analyze_size <- function(size_str) {
  convert <- if_else(str_detect(size_str,'(cm)'),1.0,2.54)
  side_a <- as.numeric(str_extract(size_str,'^(\\d+)'))*convert
  side_b <- if_else(str_detect(size_str,'(?<=\\D)(\\d+)(?=\\D)'),
                    as.numeric(str_extract(size_str,'(?<=\\D)(\\d+)(?=\\D)'))*convert, side_a)
  return(paste(sort(c(side_a, side_b, side_a*side_b)), collapse = ','))
}

picture <- read_excel(path = input, sheet = 'Pictures') %>%
  rowwise() %>%
  mutate('Side all' = analyze_size(`Size`)) %>%
  separate(`Side all`, c('Min Side', 'Max Side', 'Area'), sep=',', convert=TRUE)

frame <- read_excel(path = input, sheet = 'Frames') %>%
  rowwise() %>%
  mutate('Side all' = analyze_size(`Size`)) %>%
  separate(`Side all`, c('Frame Min Side','Frame Max Side', 'Frame Area'), sep=',', convert=TRUE) %>%
  rename('Frame' = 'Size')
  
final <- fuzzy_inner_join(picture, frame, 
                          by = c('Min Side'='Frame Min Side','Max Side'='Frame Max Side'), match_fun = list(`<=`, `<=`)) %>%
  group_by(`Picture`) %>%
  filter(`Frame Area`-`Area` == min(`Frame Area`-`Area`)) %>%
  select(c('Picture', 'Frame', 'Max Side', 'Min Side'))

View(final)
